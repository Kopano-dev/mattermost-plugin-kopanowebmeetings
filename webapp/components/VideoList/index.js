export {default, VideoList} from './VideoList.jsx';