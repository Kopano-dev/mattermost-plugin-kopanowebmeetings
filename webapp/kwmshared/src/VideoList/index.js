export {default, VideoList} from './VideoList';