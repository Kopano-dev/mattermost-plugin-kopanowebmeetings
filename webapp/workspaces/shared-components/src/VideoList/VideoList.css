.kwm-videolist {
	position: relative;
	padding: 12px;
}
.kwm-videolist video {
	background: black;
	max-height: 100%;
	max-width: 100%;
	object-fit: contain;
	position: relative;
	width: 100%;
}
.kwm-videolist .kwm-video-title {
	position: absolute;
	top: 3px;
	left: 3px;
	line-height: 1em;
	color: white;
	text-shadow: 1px 1px 0px #000;
	max-width: 60%;
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
}

.kwm-videolist .kwm-video-container {
	line-height: 0;
	overflow: hidden;
	position: relative;
}
.kwm-videolist .kwm-video-container[data-self=true] video {
	-moz-transform: scale(-1, 1);
	-ms-transform: scale(-1, 1);
	-o-transform: scale(-1, 1);
	-webkit-transform: scale(-1, 1);
	transform: scale(-1, 1);
}

