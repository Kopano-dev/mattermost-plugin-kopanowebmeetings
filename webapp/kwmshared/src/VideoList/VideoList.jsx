import React from 'react';
import PropTypes from 'prop-types';

import './VideoList.css';
import {HangupCallButton} from '../buttons';

const propTypes = {
	onHangUp: PropTypes.func,
	callers: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.string.isRequired,
		displayName: PropTypes.string.isRequired,
		stream: PropTypes.object,
		focus: PropTypes.bool,
	})),
	currentUserId: PropTypes.string.isRequired,
};
export default class VideoList extends React.Component {
	callersHash = '';

	shouldComponentUpdate(nextProps) {
		let hash = '';
		VideoList.sortCallers(nextProps.callers).forEach(caller => { // eslint-disable-line no-nested-ternary
			hash += '-' + caller.id + (caller.stream ? '1' : '0');
		});

		const shouldUpdate = hash !== this.callersHash;
		this.callersHash = hash;
		return shouldUpdate;
	}

	componentDidUpdate() {
		this.props.callers.forEach(caller => {
			const video = document.getElementById('video-' + caller.id);
			video.srcObject = caller.stream;
		});
	}

	/**
	 * Sorts the array of callers, making sure the focussed caller is listed first
	 * @param {Array} callers
	 * @return {Array} Sorted array of callers
	 */
	static sortCallers(callers) {
		return [...callers].sort((a, b) => (a.focus ? -1 : (b.focus ? 1 : 0))); // eslint-disable-line no-nested-ternary
	}

	render() {
		let buttons = '';
		if ( this.props.callers.length ) {
			buttons = (
				<div className='kwm-call-buttons'>
					<HangupCallButton onClick={this.props.onHangUp} />
				</div>
			);
		}

		return (
			<div className='kwm-videolist'>
				{VideoList.sortCallers(this.props.callers).map(caller => {
					const videoAttrs = {
						id: 'video-' + caller.id,
						autoPlay: true,
						muted: caller.id === this.props.currentUserId,
					};

					return (
						<div className='kwm-video-container' key={caller.id}>
							<video className='kwm-video' {...videoAttrs} />
							<div className='kwm-video-title'>{caller.displayName}</div>
						</div>
					);
				})}
				{buttons}
			</div>
		);
	}
}
VideoList.propTypes = propTypes;
