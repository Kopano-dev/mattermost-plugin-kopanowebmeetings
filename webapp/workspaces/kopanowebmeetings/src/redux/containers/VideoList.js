import {connect} from 'react-redux';

import * as Selectors from 'mattermost-redux/selectors/entities/users';

import {selectors} from 'redux/ducks';
const {getCallers} = selectors;
import {getDisplayName} from 'utils/utils.js';

// Original component
import VideoList from '@kopanowebmeetings/shared-components/src/VideoList';

const mapStateToProps = state => ({
	callers: getCallers(state['plugins-kopanowebmeetings']).map(c => ({
		id: c.user.id,
		displayName: getDisplayName(c.user),
		stream: c.stream,
		focus: Boolean(c.focus),
	})),
	currentUserId: Selectors.getCurrentUserId(state),
});

export default connect(mapStateToProps)(VideoList);
