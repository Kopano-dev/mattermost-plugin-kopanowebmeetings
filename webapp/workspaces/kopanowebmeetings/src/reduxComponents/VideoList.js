import {connect} from 'react-redux';

import * as Selectors from 'mattermost-redux/selectors/entities/users';

import {getDisplayName} from 'utils/utils.js';

// Original component
import VideoList from '@kopanowebmeetings/shared-components/src/VideoList';

const mapStateToProps = state => ({
	callers: state.callers.map(c => ({
		id: c.user.id,
		displayName: getDisplayName(c.user),
		stream: c.stream,
		focus: Boolean(c.focus),
	})),
	currentUserId: Selectors.getCurrentUserId(state.mattermostReduxState),
});

export default connect(mapStateToProps)(VideoList);
