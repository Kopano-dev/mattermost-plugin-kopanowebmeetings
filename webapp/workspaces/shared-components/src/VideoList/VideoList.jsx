import React from 'react';
import PropTypes from 'prop-types';

import './VideoList.css';

const propTypes = {
	callers: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.string.isRequired,
		displayName: PropTypes.string.isRequired,
		stream: PropTypes.object,
		focus: PropTypes.bool,
	})),
	currentUserId: PropTypes.string.isRequired,
	onVideoClick: PropTypes.func,
};
const defaultProps = {
	callers: [],
};
export default class VideoList extends React.Component {
	constructor(props) {
		super(props);
		this.callersHash = '';
	}

	componentDidMount() {
		this.addStreamsToVideoElements();
	}

	shouldComponentUpdate(nextProps) {
		let hash = '';
		VideoList.sortCallers(nextProps.callers, nextProps.currentUserId).forEach(caller => {
			hash += '-' + caller.id + ':' + (caller.stream ? '1' : '0') + ':' + (caller.focus ? '1' : '0');
		});

		const shouldUpdate = hash !== this.callersHash;
		this.callersHash = hash;
		return shouldUpdate;
	}

	componentDidUpdate() {
		this.addStreamsToVideoElements();
	}

	addStreamsToVideoElements() {
		this.props.callers.forEach(caller => {
			const video = document.getElementById('video-' + caller.id);
			video.srcObject = caller.stream;
		});
	}

	/**
	 * Sorts the array of callers, making sure the current user is listed first
	 * @param {Array} callers
	 * @return {Array} Sorted array of callers
	 */
	static sortCallers(callers, currentUserId) {
		return [...callers].sort((a, b) => (a.id === currentUserId ? -1 : (b.id === currentUserId ? 1 : 0))); // eslint-disable-line no-nested-ternary
	}

	render() {
		return (
			<div className='k-videolist-container'>
				<div className='kwm-videolist'>
					{VideoList.sortCallers(this.props.callers, this.props.currentUserId).map(caller => {
						const videoAttrs = {
							id: 'video-' + caller.id,
							autoPlay: true,
							muted: caller.id === this.props.currentUserId,
						};
						if ( this.props.onVideoClick ) {
							videoAttrs.onClick = e => this.props.onVideoClick(caller, e);
						}
						const displayName = caller.id === this.props.currentUserId ? 'You' : caller.displayName;

						return (
							<div className='kwm-video-container' key={caller.id} data-self={caller.id === this.props.currentUserId} data-focus={caller.focus}>
								<video className='kwm-video' {...videoAttrs} />
								<div className='kwm-video-title'>{displayName}</div>
							</div>
						);
					})}
				</div>
			</div>
		);
	}
}
VideoList.propTypes = propTypes;
VideoList.defaultProps = defaultProps;
