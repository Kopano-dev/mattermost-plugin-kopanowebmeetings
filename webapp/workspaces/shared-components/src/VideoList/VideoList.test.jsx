import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Configure Enzym to use the react-16 adapter
Enzyme.configure({adapter: new Adapter()});

import VideoList from './VideoList.jsx';

describe('VideoList component', () => {
	test('snapshot tests', () => {
		// Empty component
		let component = renderer.create(
			<VideoList
				currentUserId={'test-user-1'}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();

		// One user (never happens in reality, but should not break)
		let callers = [{
			id: 'test-user-1',
			displayName: 'Test User One',
		}];
		component = renderer.create(
			<VideoList
				currentUserId={'test-user-1'}
				callers={callers}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();

		// Two users
		callers = [{
			id: 'test-user-1',
			displayName: 'Test User One',
		}, {
			id: 'test-user-2',
			displayName: 'Test User Two',
		}];
		component = renderer.create(
			<VideoList
				currentUserId={'test-user-1'}
				callers={callers}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();

		// Five users
		callers = [{
			id: 'test-user-1',
			displayName: 'Test User One',
		}, {
			id: 'test-user-2',
			displayName: 'Test User Two',
		}, {
			id: 'test-user-3',
			displayName: 'Test User Three',
		}, {
			id: 'test-user-4',
			displayName: 'Test User Four',
		}, {
			id: 'test-user-5',
			displayName: 'Test User Five',
		}];
		component = renderer.create(
			<VideoList
				currentUserId={'test-user-1'}
				callers={callers}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	describe('shouldComponentUpdate', () => {
		test('no changes', () => {
			const callers = [{
				id: 'test-user-1',
				displayName: 'Test User One',
			}, {
				id: 'test-user-2',
				displayName: 'Test User Two',
			}];
			const wrapper = shallow(
				<VideoList
					currentUserId={'test-user-1'}
					callers={callers}
				/>
			);

			// The first time the component should be updated
			// since it hasn't been rendered yet
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(true);

			// The second time the component should not be updated
			// (because we didn't update the props)
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(false);
		});

		test('removing a user', () => {
			const callers = [{
				id: 'test-user-1',
				displayName: 'Test User One',
			}, {
				id: 'test-user-2',
				displayName: 'Test User Two',
			}];
			const wrapper = shallow(
				<VideoList
					currentUserId={'test-user-1'}
					callers={callers}
				/>
			);

			// The first time the component should be updated
			// since it hasn't been rendered yet
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(true);

			// Now remove a caller, so the component should be updated
			expect(wrapper.instance().shouldComponentUpdate({callers: callers.splice(1, 1)})).toBe(true);
		});

		test('adding a user', () => {
			const callers = [{
				id: 'test-user-1',
				displayName: 'Test User One',
			}, {
				id: 'test-user-2',
				displayName: 'Test User Two',
			}];
			const wrapper = shallow(
				<VideoList
					currentUserId={'test-user-1'}
					callers={callers}
				/>
			);

			// The first time the component should be updated
			// since it hasn't been rendered yet
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(true);

			// Now add a caller, so the component should be updated
			expect(wrapper.instance().shouldComponentUpdate({callers: callers.concat({id: 'test-user-3', displayName: 'Test User Three'})})).toBe(true);
		});

		test('changing the focussed user', () => {
			const callers = [{
				id: 'test-user-1',
				displayName: 'Test User One',
			}, {
				id: 'test-user-2',
				displayName: 'Test User Two',
			}];
			const wrapper = shallow(
				<VideoList
					currentUserId={'test-user-1'}
					callers={callers}
				/>
			);

			// The first time the component should be updated
			// since it hasn't been rendered yet
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(true);

			// Now focus the first caller, so the component should not be updated
			// (since without focus the first caller will be rendered first)
			callers[0].focus = true;
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(false);

			// Now focus the second caller, so the component should be updated
			callers[0].focus = false;
			callers[1].focus = true;
			expect(wrapper.instance().shouldComponentUpdate({callers})).toBe(true);
		});
	});

	test('sortCallers', () => {
		const callers = [{
			id: 'test-user-1',
			displayName: 'Test User One',
		}, {
			id: 'test-user-2',
			displayName: 'Test User Two',
		}, {
			id: 'test-user-3',
			displayName: 'Test User Three',
		}];

		let sortedCallers = VideoList.sortCallers(callers);
		expect(sortedCallers).toEqual(callers);

		callers[1].focus = true;
		sortedCallers = VideoList.sortCallers(callers);
		expect(sortedCallers[0]).toBe(callers[1]);

		callers[1].focus = false;
		callers[2].focus = true;
		sortedCallers = VideoList.sortCallers(callers);
		expect(sortedCallers[0]).toBe(callers[2]);
	});

	test('hangup button', () => {
		const callers = [{
			id: 'test-user-1',
			displayName: 'Test User One',
		}, {
			id: 'test-user-2',
			displayName: 'Test User Two',
		}];
		let onHangUpCallCount = 0;
		const onHangUp = () => onHangUpCallCount++;
		const component = mount(
			<VideoList
				currentUserId={callers[0].id}
				callers={callers}
				onHangUp={onHangUp}
			/>
		);
		component.find('.kwm-reject').simulate('click');
		expect(onHangUpCallCount).toBe(1);
	});
});