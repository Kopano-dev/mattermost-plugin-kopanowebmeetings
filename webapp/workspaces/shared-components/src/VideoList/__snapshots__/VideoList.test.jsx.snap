// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoList component snapshot tests 1`] = `
<div
  className="kwm-videolist"
>
  
</div>
`;

exports[`VideoList component snapshot tests 2`] = `
<div
  className="kwm-videolist"
>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-1"
      muted={true}
    />
    <div
      className="kwm-video-title"
    >
      Test User One
    </div>
  </div>
  <div
    className="kwm-call-buttons"
  >
    <button
      className="kwm-btn kwm-icon-btn kwm-reject"
      onClick={[Function]}
    >
      <svg
        className="icon-hangup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24.35,20.89a14.53,14.53,0,0,0-4.54.72v3.06a1,1,0,0,1-.56.89,12,12,0,0,0-2.63,1.83,1,1,0,0,1-.69.28,1,1,0,0,1-.7-.29l-2.44-2.44a1,1,0,0,1,0-1.41,16.81,16.81,0,0,1,23.13,0,1,1,0,0,1,.29.7,1,1,0,0,1-.29.7l-2.45,2.45a1,1,0,0,1-1.38,0,11.22,11.22,0,0,0-2.64-1.83,1,1,0,0,1-.55-.89V21.6A14.84,14.84,0,0,0,24.35,20.89Z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`VideoList component snapshot tests 3`] = `
<div
  className="kwm-videolist"
>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-1"
      muted={true}
    />
    <div
      className="kwm-video-title"
    >
      Test User One
    </div>
  </div>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-2"
      muted={false}
    />
    <div
      className="kwm-video-title"
    >
      Test User Two
    </div>
  </div>
  <div
    className="kwm-call-buttons"
  >
    <button
      className="kwm-btn kwm-icon-btn kwm-reject"
      onClick={[Function]}
    >
      <svg
        className="icon-hangup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24.35,20.89a14.53,14.53,0,0,0-4.54.72v3.06a1,1,0,0,1-.56.89,12,12,0,0,0-2.63,1.83,1,1,0,0,1-.69.28,1,1,0,0,1-.7-.29l-2.44-2.44a1,1,0,0,1,0-1.41,16.81,16.81,0,0,1,23.13,0,1,1,0,0,1,.29.7,1,1,0,0,1-.29.7l-2.45,2.45a1,1,0,0,1-1.38,0,11.22,11.22,0,0,0-2.64-1.83,1,1,0,0,1-.55-.89V21.6A14.84,14.84,0,0,0,24.35,20.89Z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`VideoList component snapshot tests 4`] = `
<div
  className="kwm-videolist"
>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-1"
      muted={true}
    />
    <div
      className="kwm-video-title"
    >
      Test User One
    </div>
  </div>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-2"
      muted={false}
    />
    <div
      className="kwm-video-title"
    >
      Test User Two
    </div>
  </div>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-3"
      muted={false}
    />
    <div
      className="kwm-video-title"
    >
      Test User Three
    </div>
  </div>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-4"
      muted={false}
    />
    <div
      className="kwm-video-title"
    >
      Test User Four
    </div>
  </div>
  <div
    className="kwm-video-container"
  >
    <video
      autoPlay={true}
      className="kwm-video"
      id="video-test-user-5"
      muted={false}
    />
    <div
      className="kwm-video-title"
    >
      Test User Five
    </div>
  </div>
  <div
    className="kwm-call-buttons"
  >
    <button
      className="kwm-btn kwm-icon-btn kwm-reject"
      onClick={[Function]}
    >
      <svg
        className="icon-hangup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24.35,20.89a14.53,14.53,0,0,0-4.54.72v3.06a1,1,0,0,1-.56.89,12,12,0,0,0-2.63,1.83,1,1,0,0,1-.69.28,1,1,0,0,1-.7-.29l-2.44-2.44a1,1,0,0,1,0-1.41,16.81,16.81,0,0,1,23.13,0,1,1,0,0,1,.29.7,1,1,0,0,1-.29.7l-2.45,2.45a1,1,0,0,1-1.38,0,11.22,11.22,0,0,0-2.64-1.83,1,1,0,0,1-.55-.89V21.6A14.84,14.84,0,0,0,24.35,20.89Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
